specification/under_development/profiles/Adaptive_Profile.md

```diff
@@ -0,0 +1,1006 @@
+  frontend](https://en.wikipedia.org/wiki/Compiler#Front_end) that produced the
+  IR to label the recorded output
+- An attribute named `"required_num_qubits"` indicating the number of qubits
+  used by the entry point
+- An attribute named `"required_num_results"` indicating the maximal number of
+  measurement results that need to be stored while executing the entry point
+  function
+
+Optionally, additional attributes may be attached to the entry point. Any custom
+function attributes attached to the entry point should be reflected as metadata
+in the program output; this includes both mandatory and optional attributes but
+not parameter attributes or return value attributes. This in particular implies
+that the [labeling schema](#output-recording) used in the recorded output can be
+identified by looking at the metadata in the produced output. See the
+specification of the [output schemas](../output_schemas/) for more information
+about how metadata is represented in the output schema.
+
+Custom function attributes will show up as part of an [attribute
+group](https://releases.llvm.org/13.0.1/docs/LangRef.html#attrgrp) in the IR.
+Attribute groups are numbered in such a way that they can be easily referenced
+by multiple function definitions or global variables. Consumers of adaptive
+profile-compliant programs must not rely on a particular numbering but instead,
+look for the function to which an attribute with the name `"entry_point"` is
+attached to determine which one to invoke when the program is launched.
+
+Both the `"entry_point"` attribute and the `"output_labeling_schema"` attribute
+can only be attached to a function definition; they are invalid on a function
+that is declared but not defined. For the Adaptive Profile, this implies that
+they can occur only in one place.
+
+Within the restrictions imposed by the Adaptive Profile, the number of qubits
+that are needed to execute the quantum program must be known at compile time.
+This number is captured in the form of the `"required_num_qubits"` attribute
+attached to the entry point. The value of the attribute must be the string
+representation of a non-negative 64-bit integer constant.
+
+Similarly, the number of measurement results that need to be stored when
+executing the entry point function is captured by the `"required_num_results"`
+attribute.
+
+Beyond the entry point-specific requirements related to attributes, custom
+attributes may optionally be attached to any of the declared functions. The
+`irreversible` attribute in particular impacts how the program logic in the
+entry point is structured, as described in the section about the [entry point
+definition](#entry-point-definition). Furthermore, the following [LLVM
+attributes](https://llvm.org/docs/LangRef.html#function-attributes) may be used
+according to their intended purpose on function declarations and call sites:
+`inlinehint`, `nofree`, `norecurse`, `readnone`, `readonly`, `writeonly`, and
+`argmemonly`.
```

Comment:
Is there a reason to highlight this specific set of LLVM attributes. I would think that if someone includes certain attributes LLVM attributes in their QIR program, consumers of that program can use some of them as hints for processing it but it would be up to them which specific attributes will have special handling.

Reply:
These seemed useful to convey information to LLVM that will enable existing optimizations. At the same time, a new/custom compiler pass should generally accurately deal with attributes, so we didn't want to give a blanket check for using any of the LLVM attributes when we defined this for the base profile. New compiler passes should respect these ones. Generally, using attributes that are not listed here usually won't result in issues, but they may not be treated/propagated properly by QIR passes.